### PR TITLE
feat(`advanced`): foundry-fork-db

### PR DESCRIPTION
**Pull request by yash-atreya**
## Motivation

The fork-db functionality was extracted from foundry as `foundry-fork-db`. This is a generic tool to create and manage forks. 

foundry's current forking mechanism and management of multiple forks is based on this. 

## Solution

Demonstrate a minimal example of how this works using `SharedBackend` and `BlockchainDb`. 

## PR Checklist

- [ ] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: examples/advanced/examples/foundry_fork_db.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    let block = BlockEnv {
+        number: U256::from(block.header.number),
+        coinbase: block.header.miner,
+        timestamp: U256::from(block.header.timestamp),
+        gas_limit: U256::from(block.header.gas_limit),
+        blob_excess_gas_and_price: None,
+        difficulty: U256::from(block.header.difficulty),
+        prevrandao: None,
+        basefee: U256::from(block.header.base_fee_per_gas.unwrap()),
+    };
```
Comment (mattsse) [inline examples/advanced/examples/foundry_fork_db.rs:40]: we should make creating BlockchainDbMeta easier with a helper function that accepts `Block`

tracking

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, this should also demonstrate how to use it with revm, like using the Database API and transacting a transfer for example

**Review by klkvr — APPROVED** [review]
nice, lgtm!

I think we should also eventually demonstrate how foundry-fork-db can be plugged into `Provider`'s `call` and `estimate_gas` fns to leverage local EVM simulation and caching of individual storage slots

**Comment by yash-atreya** [comment]
@mattsse can we have a new fork-db release and merge this?

File: examples/advanced/examples/foundry_fork_db.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    let tx_req = TransactionRequest::default()
+        .with_from(alice)
+        .with_to(bob)
+        .with_value(U256::from(100))
+        .with_max_fee_per_gas(basefee as u128)
+        .with_max_priority_fee_per_gas(basefee as u128 + 1)
+        .with_gas_limit(21000)
+        .with_nonce(0);
+
+    let mut evm = configure_evm_env(block, shared.clone(), configure_tx_env(tx_req));
```
Comment (zerosnacks) [inline examples/advanced/examples/foundry_fork_db.rs:93]: Let's say this fork environment is configured / updated on every block update, how would that look like? I think this is a common scenario.

**Review by zerosnacks — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
cool

**Pull request merged**